segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/AppendProcessor.java

```diff
@@ -385,58 +317,70 @@ private void logError(String segment, Throwable u) {
         }
     }
 
-    /**
-     * If there is too much data waiting throttle the producer by stopping consumption from the socket.
-     * If there is room for more data, we resume consuming from the socket.
-     */
-    private void pauseOrResumeReading() {
-        long bytesWaiting = getPendingBytes();
+    //endregion
 
-        if (bytesWaiting > HIGH_WATER_MARK) {
-            log.debug("Pausing writing from connection {}", connection);
-            connection.pauseReading();
+    //region WriterState
+
+    @ThreadSafe
+    private static class WriterState {
+        @Getter
+        private final Object ackLock = new Object();
+        @GuardedBy("this")
+        private long lastStoredEventNumber;
+        @GuardedBy("this")
+        private long lastAckedEventNumber;
+        @GuardedBy("this")
+        private long inFlightCount;
+
+        WriterState(long initialEventNumber) {
+            this.inFlightCount = 0;
+            this.lastStoredEventNumber = initialEventNumber;
+            this.lastAckedEventNumber = initialEventNumber;
         }
-        if (bytesWaiting < LOW_WATER_MARK) {
-            log.trace("Resuming writing from connection {}", connection);
-            connection.resumeReading();
+
+        /**
+         * Invoked when a new append is initiated.
+         *
+         * @param eventNumber The Append's Event Number.
+         * @return The previously attempted Event Number.
+         */
+        synchronized long beginAppend(long eventNumber) {
+            long previousEventNumber = this.lastStoredEventNumber;
+            Preconditions.checkState(eventNumber >= previousEventNumber, "Event was already appended.");
+            this.lastStoredEventNumber = eventNumber;
+            this.inFlightCount++;
+            return previousEventNumber;
         }
-    }
 
-    /**
-     * return the remaining bytes to be consumed.
-     */
-    protected long getPendingBytes() {
-        return pendingBytes.get();
-    }
+        /**
+         * Invoked when a conditional append has failed due to {@link BadOffsetException}. If no more appends are in the
+         * pipeline, then the Last Stored Event Number is reverted to the Last (Successfully) Acked Event Number.
```

Comment:
Shouldn't this be done unconditionally? After all won't all the intermediate writes fail with a badAttributeException?

Reply:
This is to handle offset-conditional appends, which do not close the connection. If an offset-conditional append fails due to bad offset, the WriterState's lastStoredEventNumber will have already been updated to a now-bad value; subsequent calls, even if they have the correct offset, will automatically trigger a BadAttributeUpdate and close the connection unnecessarily. This is only invoked for BadOffsetException to revert to the last known good state, which doesn't have that side effect.

This wasn't a problem before since all append were sequential and the "lastStoredEventNumber" was only updated after a successful append. This can no longer be done with pipelining; we need to update it before we make the append and deal with failure later.